JEDI based Marine DA tasks

guillaumevernieres opened the pull request:
<!-- PLEASE READ -->
<!--
Before opening a PR, please note these guidelines:

- Each PR should only address ONE topic and have an associated issue
- No hardcoded or paths to personal directories should be present
- No temporary or backup files should be committed
- Any code that was disabled by being commented out should be removed
-->

**Description**
This PR adds the relevant scripts/steps to exercise the JEDI based marine DA eventually needed for GFSv17. See issue #1072 for more details. 

The branch used for this PR was extracted from a ["rogue" branch](https://github.com/guillaumevernieres/global-workflow/tree/feature/add-soca) that included changes to the model related scripts to allow for WCDA cycling. The testing was done with the coupled `C48/5deg` atmos/ocean model, using the `GSI` for the amos initialization and the JEDI based marine DA system (SOCA) for the ocean initialization.

It is probably step 1 of at least 2 steps to get this work to a functional state in develop. I just don't know yet who's responsibility it is to fix the coupled model scripting to make it work in cycled mode.

- fixes issue #1072 (being perhaps a bit generous on my work accomplishment here ...) 

#### A few notes:
- Copying some of the `SOCA` fixed files has yet to be coordinated: An environment variable in `config.ocnanal` is pointing to `/scratch2/NCEPDEV/ocean/Guillaume.Vernieres/data/static/72x35x25/`. This needs to be copied elsewhere.
- This work as is does not cycle and will require more work on the model side as pointed above
- The testing of that ["rogue" branch](https://github.com/guillaumevernieres/global-workflow/tree/feature/add-soca) from which this PR is derived, was only done on `Hera`.


**Type of change**

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

**How Has This Been Tested?**

<!-- Please describe the tests that you ran to verify your changes and on the platforms these tests were conducted. -->
<!-- Provide instructions so we can reproduce. -->
<!-- Please also list any relevant details for your test configuration -->

<!-- Use the following as a guide to list your tests and delete options that are not relevant. Expand as necessary. -->
<!--
- [ ] Clone and Build tests on WCOSS Dell P3
- [ ] Cycled test on Orion
- [ ] Forecast-only test on Hera
-->
  
**Checklist**

- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] My changes need updates to the documentation. I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] New and existing tests pass with my changes
- [ ] Any dependent changes have been merged and published

guillaumevernieres committed:
added marine da

guillaumevernieres commented:
@aerorahul ,@WalterKolczynski-NOAA , @KateFriedman-NOAA : 
- The linter complaints seem to be outside of my code changes, do I need to address the issues?
- I don't have the necessary permissions to add reviewers, could that be changed?

aerorahul commented on sorc/link_workflow.sh:
```diff
@@ -295,10 +295,43 @@ fi
-    fv3jedi_var.x fv3jedi_convertincrement.x fv3jedi_dirac.x fv3jedi_error_covariance_training.x \
-    fv3jedi_letkf.x fv3jedi_convertstate.x fv3jedi_eda.x fv3jedi_forecast.x fv3jedi_plot_field.x \
-    fv3jedi_data_checker.py fv3jedi_enshofx.x fv3jedi_hofx_nomodel.x fv3jedi_testdata_downloader.py; do
+  declare -a JEDI_EXE=("fv3jedi_addincrement.x" \
+                       "fv3jedi_diffstates.x" \
+                       "fv3jedi_ensvariance.x" \
+                       "fv3jedi_hofx.x" \
+                       "fv3jedi_var.x" \
+                       "fv3jedi_convertincrement.x" \
+                       "fv3jedi_dirac.x" \
+                       "fv3jedi_error_covariance_training.x" \
+                       "fv3jedi_letkf.x" \
+                       "fv3jedi_convertstate.x" \
+                       "fv3jedi_eda.x" \
+                       "fv3jedi_forecast.x" \
+                       "fv3jedi_plot_field.x" \
+                       "fv3jedi_data_checker.py" \
+                       "fv3jedi_enshofx.x" \
+                       "fv3jedi_hofx_nomodel.x" \
+                       "fv3jedi_testdata_downloader.py" )
+            JEDI_EXE+=("soca_addincrement.x" \
+                        "soca_enspert.x" \
+                        "soca_hofx.x" \
+                        "soca_ensrecenter.x" \
+                        "soca_convertincrement.x" \
+                        "soca_ensvariance.x" \
+                        "soca_letkf.x" \
+                        "soca_convertstate.x" \
+                        "soca_error_covariance_training.x" \
+                        "soca_setcorscales.x" \
+                        "soca_diffstates.x" \
+                        "soca_forecast.x" \
+                        "soca_sqrtvertloc.x" \
+                        "soca_dirac.x" \
+                        "soca_gridgen.x" \
+                        "soca_staticbinit.x" \
+                        "soca_enshofx.x" \
+                        "soca_hofx3d.x" \
+                        "soca_var.x")
```
Are all of these executables used in the workflow?

guillaumevernieres replied:
@aerorahul , no, only a fraction are used currently but I'm also realizing that some of these will probably never be used. Should I minimize that list to what will eventually be used or to what's currently used?

aerorahul replied:
yes.  We should only link to what we are actively using.  There are other EE2 requirements for how they are built and the source code is linked, but I don't have an idea on how to address that issue. For e.g.`gsi.x` comes from a directory `gsi.fd`

CoryMartin-NOAA replied:
Yes this is something that we are going to have to work out with NCO, because it probably will be unreasonable to modify JEDI to put every single executable in its own folder when this is seemingly purely an aesthetic choice

aerorahul commented on workflow/rocoto/workflow_tasks.py:
```diff
@@ -578,6 +633,11 @@ def _fcst_cycled(self):
         dep = rocoto.add_dependency(dep_dict)
         dependencies = rocoto.create_dependency(dep=dep)
 
+        if self.app_config.do_jediocnvar:
+            dep_dict = {'type': 'task', 'name': f'{self.cdump}ocnanalrun'}
+            dep = rocoto.add_dependency(dep_dict)
+            dependencies = rocoto.create_dependency(dep=dep)
+
```
Does `gldas` really need to wait for ocean analysis jobs?

CoryMartin-NOAA replied:
rather instead the fcst needs to wait for ocnanal

guillaumevernieres replied:
@aerorahul : What @CoryMartin-NOAA said, forecast waiting for ocnanalrun.

aerorahul replied:
yes. I understand that dependency (ocnanalrun to complete for fcst to start).  Not sure this is going to be the case if `gldas` is turned off.

aerorahul commented on workflow/rocoto/workflow_tasks.py:
```diff
@@ -494,6 +495,60 @@ def aeroanlfinal(self):
+        deps.append(rocoto.add_dependency(dep_dict))
+        dependencies = rocoto.create_dependency(dep=deps)
+
+        cycledef = self.cdump
```
Don't think you need this.  The default for `cycledef = cdump`

aerorahul replied:
```suggestion
```

aerorahul commented on workflow/rocoto/workflow_tasks.py:
```diff
@@ -494,6 +495,60 @@ def aeroanlfinal(self):
+                              resources,
+                              cdump=self.cdump,
+                              envar=self.envars,
+                              dependency=dependencies,
+                              cycledef=cycledef)
```
```suggestion
                              dependency=dependencies)
```

aerorahul commented on parm/config/config.ocnanal:
```diff
@@ -0,0 +1,23 @@
+export R2D2_OBS_DUMP=s2s_v1
+export R2D2_OBS_SRC=gdas_marine
+export R2D2_OBS_WINDOW=24     # TODO: Check if the R2D2 sampling DB window is still needed
+export COMIN_OBS=/scratch2/NCEPDEV/marineda/r2d2
```
Perhaps add a `TODO` here to remind us to move this to a common space.

guillaumevernieres replied:
@aerorahul : A common space for R2D2 stuff or for the entirety of config.ocnanal?

aerorahul replied:
for R2D2

aerorahul commented on parm/config/config.ocnanal:
```diff
@@ -0,0 +1,23 @@
+#!/bin/bash -x
```
I don't think we need a `-x` here.

aerorahul reviewed (approved):
Changes look good.
A few comments and a review from the other JEDI developers would be helpful.

CoryMartin-NOAA commented on jobs/rocoto/ocnanalpost.sh:
```diff
@@ -0,0 +1,15 @@
+# Source GDASApp modules
+module purge
+module use "${HOMEgfs}"/sorc/gdas.cd/modulefiles
+module load GDAS/hera
```
This shouldn't be machine dependent here.

CoryMartin-NOAA commented on jobs/rocoto/ocnanalprep.sh:
```diff
@@ -0,0 +1,15 @@
+# Source GDASApp modules
+module purge
+module use "${HOMEgfs}"/sorc/gdas.cd/modulefiles
+module load GDAS/hera
```
same comment

CoryMartin-NOAA commented on jobs/rocoto/ocnanalrun.sh:
```diff
@@ -0,0 +1,15 @@
+# Source GDASApp modules
+module purge
+module use "${HOMEgfs}"/sorc/gdas.cd/modulefiles
+module load GDAS/hera
```
x3

CoryMartin-NOAA commented on parm/config/config.resources:
```diff
@@ -156,77 +156,128 @@ elif [ ${step} = "waveawipsgridded" ]; then
+        npes=16
+        ;;
+      *)
+          echo "FATAL: Resolution not supported'"
```
This will be a problem eventually but fine for now

CoryMartin-NOAA reviewed (commented):
A few comments, mostly looks good to me

guillaumevernieres committed:
Update workflow/rocoto/workflow_tasks.py

Co-authored-by: Rahul Mahajan <aerorahul@users.noreply.github.com>

guillaumevernieres committed:
Update workflow/rocoto/workflow_tasks.py

Co-authored-by: Rahul Mahajan <aerorahul@users.noreply.github.com>

guillaumevernieres committed:
addressed reviewer's comments

guillaumevernieres committed:
cn

guillaumevernieres commented:
Thanks for your reviews/suggested commits @aerorahul and @CoryMartin-NOAA . I've addressed your comments and merged your suggestions.

aerorahul commented on jobs/JGDAS_GLOBAL_OCEAN_ANALYSIS_POST:
```diff
@@ -0,0 +1,23 @@
+    [[ "${status}" -ne 0 ]] && exit "${status}"
+done
+
+##########################################
+# Remove the Temporary working directory
+##########################################
+cd "${DATAROOT}" || exit 1
+[[ ${KEEPDATA} = "NO" ]] && rm -rf "${DATA}"
```
This job is doing nothing 🥇

guillaumevernieres replied:
Thanks for the medal @aerorahul .

I assume empty stubs are fine? This will be populated in a subsequent pr, but I currently have no post processing step.

aerorahul commented on jobs/JGDAS_GLOBAL_OCEAN_ANALYSIS_RUN:
```diff
@@ -0,0 +1,94 @@
+
+# Prepare soca static dir
+
+ln -s "${COMOUT}"/analysis/* .
```
Why is this in a j-job?  @CoryMartin-NOAA

guillaumevernieres replied:
Thanks for putting the burden of explaining what I did on @CoryMartin-NOAA ! So, Cory, what's up with that?

guillaumevernieres replied:
@aerorahul I remember what the issue was. All the yaml's , static files, backgrounds, obs, etc ... are created/linked/copied in the "prep" step under ```${COMOUT}"/analysis/``` and the content of that directory is needed by the run step.

CoryMartin-NOAA replied:
@guillaumevernieres follow what I have here for the aero case:
```
export DATA=${DATA:-${DATAROOT}/aeroanl_${CDATE}}
mkdir -p "${DATA}"
cd "${DATA}" || (echo "${DATA} does not exist. ABORT!"; exit 1)
```
https://github.com/NOAA-EMC/global-workflow/blob/1536029cd857585af626147c36f3a053b223e11a/jobs/JGDAS_GLOBAL_AERO_ANALYSIS_FINALIZE#L48-L50

guillaumevernieres committed:
prep/run in same scratch

guillaumevernieres commented:
@aerorahul , @CoryMartin-NOAA , I just pushed a commit that addresses most of your comments except that I'm still not doing anything in `post`. I'll have a separate PR that will populate `post` if it's OK with you.

aerorahul commented on jobs/JGDAS_GLOBAL_OCEAN_ANALYSIS_PREP:
```diff
@@ -0,0 +1,107 @@
+##########################################
+# Remove the Temporary working directory
+##########################################
+cd "${DATAROOT}" || exit 1
+[[ ${KEEPDATA} = "NO" ]] && rm -rf "${DATA}"
```
if you remove this here, you won't be able to use it in the `ocnanalrun` job.
I think you need to keep this.

guillaumevernieres replied:
Oops, thanks ... I just removed these lines.

guillaumevernieres committed:
keep DATA for run step

CoryMartin-NOAA commented on jobs/JGDAS_GLOBAL_OCEAN_ANALYSIS_RUN:
```diff
@@ -0,0 +1,90 @@
+# Remove the Temporary working directory
+##########################################
+cd "${DATAROOT}" || exit 1
+[[ ${KEEPDATA} = "NO" ]] && rm -rf "${DATA}"
```
Don't want this one either I don't think unless you have nothing to archive/save from this job back to the ROTDIR.

guillaumevernieres replied:
@CoryMartin-NOAA , unless I messed up (very likely), the analysis and increment are dumped in `COMOUT` (==`ROTDIR`?)  ... And as I write this, I realize I forgot to do the same for `diag`. Is it OK to have symlinks to directories in `COMOUT`/`ROTDIR` in which we directly write what we want to keep?

CoryMartin-NOAA replied:
for now it is, but 'eventually' it is not. NCO wants no symlinks, and it makes sense from a solid state drive for run / spinning disk for storage perspective. What my plan is/was for the aerosol tasks (see PR#1106) is to copy everything I want to save from  $DATA to $COMOUT as part of the finalize job after the run job.

guillaumevernieres replied:
Sounds good @CoryMartin-NOAA , I'll do this in a subsequent PR if that's OK with you (I'm just being polite and respectful in exchange for a green check). My next step is to reconcile my cycling branch with the changes here and make it work again, which I'm pretty sure will require some adjustment/bug fixes in the GDASApp and the g-w. Looking forward to it!

CoryMartin-NOAA commented on sorc/link_workflow.sh:
```diff
@@ -295,10 +295,30 @@ fi
+                       "fv3jedi_data_checker.py" \
+                       "fv3jedi_enshofx.x" \
+                       "fv3jedi_hofx_nomodel.x" \
+                       "fv3jedi_testdata_downloader.py" )
```
Note to myself and @RussTreadon-NOAA we will also want to prune this list for fv3jedi executables at a later date

CoryMartin-NOAA reviewed (commented):
one todo for me, one thing that needs fixed by @guillaumevernieres

CoryMartin-NOAA reviewed (approved):
Approved pending conflict resolutions

guillaumevernieres committed:
Merge branch 'develop' into feature/marine_gdas

WalterKolczynski-NOAA commented on jobs/rocoto/ocnanalpost.sh:
```diff
@@ -0,0 +1,16 @@
+
+###############################################################
+# Source GDASApp modules
+declare -l hpc=${machine}
```
Since `hpc` is only used once, it would probably be better to just use the bash lower-case substitution `${machine,,}` inline on L10 instead of declaring an extra variable.

WalterKolczynski-NOAA commented on jobs/rocoto/ocnanalpost.sh:
```diff
@@ -0,0 +1,16 @@
+#! /usr/bin/env bash
+
+source "${HOMEgfs}"/ush/preamble.sh
```
I know this is how shellcheck "fixes" it, but it is preferred to quote the entire path. (Several more in this file not noted.)

WalterKolczynski-NOAA commented on jobs/rocoto/ocnanalprep.sh:
```diff
@@ -0,0 +1,16 @@
+#! /usr/bin/env bash
```
Same corrections as for `ocnanalpost.sh`.

WalterKolczynski-NOAA commented on jobs/rocoto/ocnanalrun.sh:
```diff
@@ -0,0 +1,16 @@
+#! /usr/bin/env bash
```
Same as before.

WalterKolczynski-NOAA commented on parm/config/config.ocnanalprep:
```diff
@@ -0,0 +1,10 @@
+#!/bin/bash -x
```
Other config files aren't turning off trace on their own, so this one shouldn't either.

WalterKolczynski-NOAA commented on parm/config/config.ocnanalrun:
```diff
@@ -0,0 +1,11 @@
+#!/bin/bash -x
```
Other config files aren't turning off trace on their own, so this one shouldn't either.

WalterKolczynski-NOAA commented on parm/config/config.ocnanalpost:
```diff
@@ -0,0 +1,10 @@
+#!/bin/bash -x
```
Other config files aren't turning off trace on their own, so this one shouldn't either.

WalterKolczynski-NOAA commented on sorc/link_workflow.sh:
```diff
@@ -295,10 +295,30 @@ fi
+                       "fv3jedi_enshofx.x" \
+                       "fv3jedi_hofx_nomodel.x" \
+                       "fv3jedi_testdata_downloader.py" )
+            JEDI_EXE+=(
```
Fine with putting these in a list (we should probably do the same for the others in this file), but why two separate declarations without any control structures making the appended bit conditional?

guillaumevernieres replied:
I don't know, to remind the reviewer to inhale in the middle of reading the list? Removed.

WalterKolczynski-NOAA reviewed (changes requested):
Develop has been updated to eliminate all of the existing shellcheck issues in the `*.env` files, so this PR will need to address all those present in the new code as well.

guillaumevernieres committed:
tidy code according to rev comments

guillaumevernieres commented:
Thanks for your review @WalterKolczynski-NOAA , I've addressed your comments.
